server/container_create_linux.go

```diff
@@ -229,6 +231,33 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Cont
 		someRepoDigest = imgResult.RepoDigests[0]
 	}
 
+	systemCtx, err := s.contextForNamespace(sb.Metadata().Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("get context for namespace: %w", err)
+	}
+
+	userSpecifiedImage := ctr.Config().GetImage().UserSpecifiedImage
+
+	if userSpecifiedImage == "" {
+		// Attempt to check if the image is a checkpoint OCI image
+		if _, err := s.checkIfCheckpointOCIImage(ctx, userRequestedImage); err != nil {
+			return nil, fmt.Errorf("failed to check if this is a checkpoint image: %w", err)
+		} else { // Fallback to ensure that userSpecifiedImage is valid
+			userSpecifiedImage = imageName.StringForOutOfProcessConsumptionOnly()
+		}
+	}
+
+	var userImageRef references.RegistryImageReference
+
+	userImageRef, err = references.ParseRegistryImageReferenceFromOutOfProcessData(userSpecifiedImage)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get userImageRef from user specified image %q: %w", userSpecifiedImage, err)
+	}
+
+	if err := s.StorageImageServer().IsRunningImageAllowed(ctx, &systemCtx, userImageRef, imgResult.Digest); err != nil {
```

Comment:
Using `imgResult.Digest` here is not a _security_ issue _yet_ but seems incorrect. `.Digest` is _some_ digest of this deduplicated image, not necessarily the one the user wanted or one that is signed.

It is possible to have a single image deduplicated image for `repo1:tag1 == digest1` and `repo2:tag2 == digest2`. And we can have `UserSpecifiedImage` (user intent) for repo1:tag1, but `imgResult` returning `digest2`. The way `IsRunningImageAllowed` is implemented, this will check the `digest2` manifest for signatures, and even if the image is correctly signed for `repo1:tag1`, the signature verification will fail. (To make that example even more vivid, consider user input `repo1@digest1`, with `imgResult.Digest == digest2`. That should intuitively fail any policy check.)

---

⚠️ 

This _will_ become a security issue in a bit: Currently the same image digest implies the same image ID, but that’s no longer the case with zstd:chunked (where it matters _how_ we pulled the image).

Basically, the “which image are we verifying” parameter to `IsRunningImageAllowed` should be `imageID`, and the `imageID` should participate in `istorage.Transport.NewStoreReference`.

Then, in the `IsRunningImage` implementation, we can discuss the tradeoffs of calling `NewStoreReference` _only_ with the image ID vs. also with `userSpecifiedImageRef`, or possibly trying both.

Reply:
updated to use the imageID, which simplifies things anyway. Because we already have access to imageRef, I chose not to use userSpecifiedImageRef there.